### DISCUSSION TITLE
Confusion about dev dependencies - what controls them?

### DISCUSSION THREAD

**Discussion by Destroy666x** [pdm-project/pdm#2289] [Q&A]
Hello,

I'm probably missing something obvious, but using pdm for the 1st time and I'm having a hard time figuring out how to install dev dependencies. I have this section:
```
[tool.pdm.dev-dependencies]
style = [
  'black==2.9.*',
  'ruff==0.0.*',
]

test = [
  'pytest==7.4.*',
]
```

After reading the documentation, I'd assume that these will be installed by default with `pdm install`. But they aren't. And `pdm install -d` reports:
```
[PdmUsageError]: Requested groups not in lockfile: test,style
```
which doesn't quite make sense to me as the install command is supposed to modify the lock file.

What am I missing? I'm using pdm 2.9.3.

**Comment by sr-murthy** [comment]
`pdm install` installs the dependencies specified in a lockfile - either the default root-level lockfile, if no `-L` option is given, or any lockfile specified with `-L`.

If you leave out `-L /path/to/lockfile` then `pdm install` will use the default lockfile, `pdm.lock`, in your root directory, and in your case it looks like your default lockfile does not contain those dev dependencies in `pyproject.toml`. 

I suppose one way of syncing the default lockfile with `pyproject.toml` is to run `pdm update -d`, which will result in possible package updates in `pyproject.toml` in both the default (production) and dev. dependencies, if newer versions were detected upstream, as well as resulting changes to the lockfile and your working set. So in your case I would first run:

    pdm update -d

As this would end up installing any new or updated dependencies in your working environment from an updated lockfile, a subsequent `pdm install` wouldn't be necessary. 

So, basically your lockfile(s) should be in sync with your `pyproject.toml`, and then `pdm install` would work.

**Reply by Destroy666x** [reply]
I see, thanks, so that's more in-line with some other languages' dependency managers IIRC (although it differs from the ones I know the most). The documentation had me completely confused there.

Now a question - how to install new dev dependencies provided through toml (not CLI) without updating other libraries? Is there any possibility like this? I noticed that freshly added regular non-dev packages get installed with just `pdm install`, which also added to confusion.

**Reply by Destroy666x** [reply]
E.g. here I added `requests` to regular dependencies, ran `pdm install` and that got installed despite not existing in lock with `Lock file hash doesn't match pyproject.toml, packages may be outdated. Updating the lock file...` part:
![image](https://github.com/pdm-project/pdm/assets/3606072/85ad4af3-fbde-403f-a6df-8e2ef6351555) 

Yet dev dependencies were left in tact and lockfile wasn't updated for them.

**Reply by frostming** [reply]
>ran pdm install and that got installed despite not existing in lock

That's the purpose of pdm install, to keep the lockfile and pyproject and working set in sync. And since the newly added group was not in the lockfile it won't be locked nor installed

**Reply by sr-murthy** [reply]
@Destroy666x You're not limited to just one lockfile. You can create and maintain multiple lockfiles, e.g. a default lockfile for prod. dependencies (and sub-dependencies), or a dev. one with all dev. dependencies (and sub-dependencies), including the prod. dependencies, or a test dependencies lockfile etc. And you can specify these with `pdm install`. Obviously, you can only specify one lockfile at a time, and your environment can only be synced with one lockfile at a time.

Note that, like `pdm add` and `pdm update`, the `pdm install` command does have a `--no-lock` option in case you don't want a lockfile to reflect changes from the command. In my case, this also depends on version control: in my main work project, which uses PDM, I only version control the default root lockfile, which contains prod. dependencies only. I have a separate dev lockfile which I do not version control.

**Reply by sr-murthy** [reply]
@Destroy666x If you want to add dependencies (in any group, either prod. or dev.) then you can use `pdm add`:

https://pdm.fming.dev/latest/reference/cli/#add

This will add the new dependencies to the relevant groups in `pyproject.toml`, and then assuming you've left out the `--no-lock` and `--no-sync` options, it will also update the (default) lockfile and install them in your environment. It will of course install all sub-dependencies of any new dependency. 

So instead of changing `pyproject.toml` manually, use `pdm add`, `pdm update`, `pdm remove` etc. Any resulting changes to `pyproject.toml` and the default lockfile, or any actually any lockfile you're version controlling, should be staged and committed, obviously.

**Reply by Destroy666x** [reply]
Hmm, I just tried mentioned `pdm update -d`, even though updating libs wouldn't be desired, but doesn't matter for now, and that also complained that `style` and `test` groups aren't in the lock file.

@sr-murthy so no way to do it without using CLI `add` you're saying?

**Reply by sr-murthy** [reply]
@Destroy666x Your lockfile is not consistent with `pyproject.toml`. Can you please try undoing those manual additions to `pyproject.toml`, and then try adding them with `pdm add` instead? So, first, revert your changes to `pyproject.toml`. Then, assuming your default lockfile is `pdm.lock` and in the project root, run `pdm add` to add `style` to `pyproject.toml` + update lockfile + sync with env.:

    pdm add -dG "style" black ruff

This will add this group and those dependencies to `pyproject.toml`, so changes to this file should be staged and committed if you don't want to lose them. But this will also update the default lockfile (as we've left out `-L` and also left out `--no-lock`) with those dependencies, and install them in the environment (as we've left out `--no-sync`). Now your lockfile and working set should be in sync with `pyproject.toml`, and if for any reason you need to re-install them `pdm install` should work, because those dependencies would already be in the lockfile.

Which lockfile you use for dev dependencies, and whether you want to use the default or a separate one is up to you.

**Reply by sr-murthy** [reply]
To install `requests` as a production dependency try running this:

    pdm add --no-editable requests

This will add `requests` to the prod. `dependencies` group in `pyproject.toml`, and update the default lockfile (we've left out `-L` and `--no-lock`) and also add it to the working set (we've left out `--no-sync`). I've used `--no-editable` as I'm assuming you want a non-editable version of `requests`, but I've also left out `--no-self` as I'm assuming you're working in editable/development mode. So leavjng out `--no-self` means it will install the project in development mode. If you don't want this include `--no-self`.

If you want this change in `pyproject.toml` then you should obviously stage and commit this file, and also `pdm.lock`. On mine these are the changes it produces:

     $ git diff pyproject.toml
     diff --git a/pyproject.toml b/pyproject.toml
     index e125d434..8cb7adc1 100644
     --- a/pyproject.toml
     +++ b/pyproject.toml
     @@ -30,6 +30,7 @@ dependencies = [
          "tomli>=1.1.0; python_version < \"3.11\"",
          "importlib-resources>=5; python_version < \"3.9\"",
          "importlib-metadata>=3.6; python_version < \"3.10\"",
     +    "requests>=2.31.0",
      ]
      readme = "README.md"
      keywords = ["packaging", "dependency", "workflow"]
     (pdm-env) [2023-10-04 08:31:34 srm@~/Documents/dev/pdm]$ git diff pdm.lock
     diff --git a/pdm.lock b/pdm.lock
     index bb8a246b..e25abbbf 100644
     --- a/pdm.lock
     +++ b/pdm.lock
     @@ -6,7 +6,7 @@ groups = ["default", "all", "doc", "pytest", "test", "tox", "workflow"]
      cross_platform = true
      static_urls = false
      lock_version = "4.3"
     -content_hash = "sha256:a981bd6e57936e05ef989e690cff324b4e3c88858f026db22afc1caa35013ff8"
     +content_hash = "sha256:9fa79d1e013c54883edc66b817316d3d0641f6fd7f82f35ca36b781eaf486bf4"
      
      [[package]]
      name = "arpeggio"

**Reply by Destroy666x** [reply]
Sorry, but this is getting more and more confusing with each reply.

As I said, `requests` had no issue with that flow. It got automatically installed with `pdm install` after being added to `pyproject.toml`. So do any other libraries added directly to `dependencies`. So no issue with that and not sure why you wrote the comment above.

The only issue I had was with dev dependencies defined separately under own section defined by this tool - `tool.pdm.dev-dependencies`, which are inconsistent with that flow. Thanks for explaining the `add` command, I'll switch to that if I'll have to. However, my main question I asked above stays unanswered - is there a way to do it without installing each dev package separately with `add` in CLI? As it's usually faster to edit the file directly with new packages in some other dependency managers for different programming languages. I assume that was a "no", but I'd like a direct answer, please.

**Reply by sr-murthy** [reply]
@Destroy666x Firstly, if the `requests`step wasn't required then you can ignore that. The answer to your last question - on `install` of dev deps - is "no", because as the install command help **clearly states**:

    Install dependencies from lock file

Is that confusing? You manually updated `pyproject.toml` to add new dev dependencies, which immediately makes your lockfile(s) inconsistent. If you're adding, updating or removing deps of any type (prod or dev) use the `add`, `update`, or `remove` commands in the right way. I see `install` as a step for re-installing any deps that I may accidentally removed from my env., and `install` works from a lockfile, not directly from `pyproject.toml`.

**Reply by Destroy666x** [reply]
Since it **doesn't** install only dependencies from lock file, that aren't dev (or grouped) dependencies, like I mentioned - yes, I'd definitely say it's confusing (and I see not only for me). Pretty sure it's more related to **groups** that @frostming mentioned but that's not documented too well IMO.

Now I did a quick test to confirm - once you run `pdm add -d -G style black`, then the command states `Adding group style to lockfile` and anything added to `style` is installed... E.g. 
```
[tool.pdm.dev-dependencies]
style = [
    "black==23.9.*",
    'ruff==0.0.*',
    'autopep8==2.0.*',
]
```
installed `ruff` and `autopep8` despite running only the command above. So lock doesn't lock dependencies for install command at all and stating that is wrong, it locks **groups of dependencies**, for some reason.

Now I'll have to scan the documentation to see if there's a way to add/lock a group instead of adding/locking a dependency with a group. EDIT: yeah, can just run `pdm add -d -G somegroup` and then install.

**Reply by sr-murthy** [reply]
@Destroy666x I've only been using PDM for four months, and I haven't used `install` a lot, except when I installed my project dependencies for the first time, after I had first installed PDM from a preexisting `pyproject.toml`. When you have a preexisting project TOML and then install PDM, it will generate a lockfile for you, and that will only contain prod. dependencies. Then I remember running `pdm install --dev` to install all deps and the project in editable mode, but used `--no-lock` to to avoid changing the default lockfile. I version control the default lockfile, and only update it via `add` / `update` / `remove`. I keep a separate lockfile for dev deps which I do not version control.

Since the initial setup I now avoid `install` and use `add`/`update`/`remove` instead as I want deps changes to be reflected in the TOML file and lockfile and env. simultaneously.

I don't know much about `lock` as I haven't used it, and I'm still working my way through the CLI and the codebase. But from what you say it 's possible that the some CLI docs may need to be updated

**Reply by Destroy666x** [reply]
Not sure what you mean with dependency part, but `autopep8` is surely not a dependency of `black` - they're 2 different formatters: https://github.com/psf/black/blob/main/pyproject.toml#L64-L72 I think you Ctrl + Fd to `keywords`?

**Reply by sr-murthy** [reply]
Ah OK, yes, you're right.

**Reply by sr-murthy** [reply]
@Destroy666x I thought I would test your proposition that `pdm add -dG "style" black` adds `ruff` and `autopep8`. It did not:

    $ pdm add -dG "style" black
    ...
    
    Adding group style to lockfile
    Adding packages to style dev-dependencies: black
    🔒 Lock successful
    Changes are written to pyproject.toml.
    Synchronizing working set with resolved packages: 2 to add, 0 to update, 0 to remove

      ✔ Install mypy-extensions 1.0.0 successful
      ✔ Install black 23.3.0 successful

    🎉 All complete!

Indeed, `pyproject.toml` only contains a single change: a new `style` dev dep group with `black` added:

    +style = [
    +        "black>=23.3.0",
    +] 
    +

(To undo I ran  `pdm remove -dG "style" black`.

**Reply by Destroy666x** [reply]
Did you have the other packages listed in pyproject.toml in style group? That diff states otherwise

**Reply by sr-murthy** [reply]
No @Destroy666x I did not have a `style` group prior to running this command. The diff I posted is what I'm seeing and does not suggest otherwise: the `style` group with `black` was added to dev deps, and that's it.

**Reply by sr-murthy** [reply]
You did not post the full output of your commands. Please do that, and then we can verify your assertions.

**Reply by Destroy666x** [reply]
I listed the command and what I had in pyproject.toml before I ran it, then I wrote what the result was. There's nothing else involved. I'll leave it at that and will wait for tool's author confirmation if this is how this works.

**Reply by sr-murthy** [reply]
Sure, but I was just trying to reproduce your error, which I could not.

**Comment by sr-murthy** [comment]
@frostming This raises a question. We have a `pdm sync` that syncs the working set with the lockfile, but do we need to extend this to `pyproject.toml` as well, so syncing between `pyproject.toml` <-> lockfile <-> working set?

**Reply by frostming** [reply]
> @frostming This raises a question. We have a `pdm sync` that syncs the working set with the lockfile, but do we need to extend this to `pyproject.toml` as well, so syncing between `pyproject.toml` <-> lockfile <-> working set?
> 
> 

that's what pdm install does
